text/0000-forbidden-value-trait.md

```diff
@@ -0,0 +1,70 @@
+- Start Date: 2020-03-24
+- RFC PR #:
+- Rust Issue #:
+
+# Summary
+
+Some types have one or more forbidden values. Such values can be sometimes used to perform [optimizations](https://github.com/rust-lang/rust/pull/45225). For example, rustc currently uses a known forbidden value of a certain type `A` (like `&T`, `Box<T>`, `NonNull<T>`, `NonZeroI8` and some enum types) to represent the `None` enum variant of `Option<A>`. I propose to give users the ability to enable the same optimizations for types rustc currently doesn't know a forbidden value exists for.
+
+# Motivation
+
+This would increase the number of situations where the compiler can do the aforementioned, genuinely useful optimizations.
+
+# Detailed design
+
+Add the following trait somewhere in the standard library:
+```rust
+unsafe trait ForbiddenValues<A, B>
+where
+    A: FixedSizeArray<B>,
+    B: FixedSizeArray<u8>,
+{
+    const FORBIDDEN_VALUES: A;
+}
+```
+
+To implement `ForbiddenValue<A, B>` for type `T`, the following conditions must be met:
+1) Type `T` has a stable layout and representation
+2) Type `B` is a fixed-size array of `u8`
+3) Type `A` is a fixed-size array of `B`
+4) `std::mem::size_of::<T>() == std::mem::size_of::<B>()`
+5) For each value `v` of type `T`, and for each item `f` in `T::FORBIDDEN_VALUES`, `*(&v as *const T as *const B) != f`
+
+Then compilers would be allowed to use `T::FORBIDDEN_VALUES` to represent forbidden values of type `T` in whatever optimizations they decide to perform.
+
+Users could then, for example, implement a wrapper for floating point values that are always finite like this:
+```rust
+#![feature(const_transmute)]
+
+#[repr(transparent)]
+struct FastFloat(f32);
+
+unsafe impl ForbiddenValues<[[u8; 4]; 3], [u8; 4]> for FastFloat {
+    const FORBIDDEN_VALUES: [[u8; 4]; 3] = unsafe {
+        [
+            mem::transmute(f32::NAN),
+            mem::transmute(f32::INFINITY),
+            mem::transmute(f32::NEG_INFINITY),
+        ]
```

Comment:
How do we have a range of forbidden values? For example if we were to have `u8` that is able to store 55 numbers at most, then all the rest of the numbers as forbidden values.

Reply:
You would have to add all those 201 (2^8-55) forbidden values to the `FORBIDDEN_VALUES` array as `[u8; 1]` values.

Reply:
That seems extremely suboptimal. Why not just accept ranges? Individual values can be done as ranges too.

Reply:
I have a struct that is packed and uses 30 bits. For obvious reasons it's stored as a 32 bit integer.

If we used `const fn(T) -> bool` or ranges, this would be trivial to write. With the proposal, I'd have to list out _over a billion values_. Obviously that's neither feasible nor fast.

Reply:
I agree that a list is infeasible, but my point is that a predicate is also infeasible. Imagine that the forbidden values are in the higher bits, and that the compiler simply does a sweep of the values in order to find the forbidden ones. It'll still have to go through almost anything.

And if your argument is that the compiler should be able to look at the function and determine what it's doing, 1. the halting problem and 2. ranges are easier and way less jank :p

Reply:
Slice of ranges? I think that would solve pretty much every situation, as a value can (of course) be represented as a range.

Reply:
@clarfon Do you mean ranges as in something like the following?
```rust
unsafe trait ForbiddenValues<A, B>
where
    A: FixedSizeArray<(B, B)>,
    B: FixedSizeArray<u8>,
{
    const FORBIDDEN_VALUES: A;
}
```
Where `B` is now interpreted as a fixed-size unsigned integer type with an arbitrary width, and the tuple `(B, B)` represents the inclusive start and end of the range of forbidden values.

Reply:
I think using a list of both ranges and bit masks to define forbidden values is probably the best solution, since it handles nearly all cases and doesn't have the halting problem and similar issues you'd run into with defining it using a `const fn`.

For the bit mask, it would match all values `v` that pass `(v & a) == b` where `a` and `b` are values of some sort (byte arrays? integers?).

As an example of why ranges alone are insufficient:
What if you have a type like:
```rust
enum E<T> {
    A(T),
    B(T),
    C(T),
    D(T),
}

type E2<'a'> = E<&'a u32>;
```

where you want Rust to pack the enum tag for `E2` into the 2 low bits of the pointer since the pointer is guaranteed to be some value with the two low bits are zeros due to references guaranteed alignment and `u32` having an alignment of 4? Rust currently doesn't pack the enum that way, but could. This kind of bit packing is used extensively throughout LLVM's codebase (implemented using custom C++ code), so is not that uncommon.

If you were to enumerate the forbidden values of a custom `T` type using ranges, you'd get something like:
```rust
const FORBIDDEN_VALUES: [...] = [
    0x0..=0x3, // include 0 for null
    0x5..=0x7, // repeats every 4 values
    0x9..=0xB,
    0xD..=0xF,
    0x11..=0x13,
    0x15..=0x17,
    0x19..=0x1B,
    0x1D..=0x1F,
    ... // another *billion* lines for 32-bit pointers
];
```

Reply:
Something like the following to allow forbidding values both by ranges and by a bit mask?
```rust
unsafe trait ForbiddenValues<R, B>
where
    R: FixedSizeArray<RangeInclusive<B>>,
    B: FixedSizeArray<u8>,
{
    const FORBIDDEN_RANGES: R;
    const FORBIDDEN_BITS: B;
}
```

Reply:
For bitmasks, just a single integer is insufficient, you'd need something like:

```rust
struct BitMask<T> {
    mask: T,
    value: T,
}

impl<T> BitMask<T> {
    fn matches<'a>(&'a self, value: T) -> bool
    where
        T: BitAnd<&'a T, Output = T> + PartialEq,
    {
        (value & &self.mask) == self.value
    }
}

unsafe trait ForbiddenValues<B>
where
    B: MemoryBits,
{
    const FORBIDDEN_RANGES: &'static [RangeInclusive<B>];
    const FORBIDDEN_BIT_MASKS: &'static [BitMask<B>];
}
```

This allows you to represent things such as an integer where top 4 and bottom 12 bits can't be `0x3` and `0xCAB` by including a bit mask `BitMask { mask: 0xF000_0FFFu32, value: 0x3000_0CABu32 }` assuming `u32` implements `MemoryBits`.